common/src/main/java/io/homeassistant/companion/android/common/data/prefs/WearPrefsRepositoryImpl.kt

```diff
@@ -52,15 +54,64 @@ class WearPrefsRepositoryImpl @Inject constructor(
         }
     }
 
-    override suspend fun getTileShortcuts(): List<String> {
-        val jsonArray = JSONArray(localStorage.getString(PREF_TILE_SHORTCUTS) ?: "[]")
-        return List(jsonArray.length()) {
-            jsonArray.getString(it)
-        }
+    override suspend fun getTileShortcuts(tileId: Int): List<String> =
+        localStorage.getString(PREF_TILE_SHORTCUTS)?.let { jsonStr ->
+            runCatching {
+                val jsonObject = JSONObject(jsonStr)
+                val key = tileId.toString()
+                if (jsonObject.has(key)) {
+                    val jsonArray = jsonObject.getJSONArray(key)
+                    jsonArray.toStringList()
+                } else {
+                    null
+                }
+            }.recover {
+                // backwards compatibility with the previous format when there was only one Shortcut Tile:
+                val jsonArray = JSONArray(jsonStr)
+                val entities = jsonArray.toStringList()
+                setTileShortcuts(mapOf(tileId to entities))
+                entities
+            }.getOrNull()
+        } ?: emptyList()
+
+    override suspend fun getAllTileShortcuts(): Map<out Int?, List<String>> {
+        return localStorage.getString(PREF_TILE_SHORTCUTS)?.let { jsonStr ->
+            runCatching {
+                val jsonObject = JSONObject(jsonStr)
+                buildMap {
+                    jsonObject.keys().forEach { key ->
+                        val intKey = key.toInt()
+                        val jsonArray = jsonObject.getJSONArray(key)
+                        val entities = jsonArray.toStringList()
+                        put(intKey, entities)
+                    }
+                }
+            }.recover {
+                // backwards compatibility with the previous format when there was only one Shortcut Tile:
+                val jsonArray = JSONArray(jsonStr)
```

Comment:
I had one tile of shortcuts before installing this, an added a second tile after that resulting in the following data for my preference: `{&quot;null&quot;:[&quot;switch.stekkerblok,Stekkerblok,&quot;,&quot;switch.vaatwasser,Vaatwasser,mdi:dishwasher&quot;],&quot;18&quot;:[]}` (directly copied from the wear_0.xml file, not completely sure if the `&quot;` is Android Studio doing conversions or not)

This now causes the app to crash on launch with:
```
2023-07-28 16:08:46.187  7384-7384  AndroidRuntime          io....stant.companion.android.debug  E  FATAL EXCEPTION: main
                                                                                                    Process: io.homeassistant.companion.android.debug, PID: 7384
                                                                                                    org.json.JSONException: Value {"null":["switch.stekkerblok,Stekkerblok,","switch.vaatwasser,Vaatwasser,mdi:dishwasher"],"18":[]} of type org.json.JSONObject cannot be converted to JSONArray
                                                                                                    	at org.json.JSON.typeMismatch(JSON.java:112)
                                                                                                    	at org.json.JSONArray.<init>(JSONArray.java:98)
                                                                                                    	at org.json.JSONArray.<init>(JSONArray.java:110)
                                                                                                    	at io.homeassistant.companion.android.common.data.prefs.WearPrefsRepositoryImpl.getAllTileShortcuts(WearPrefsRepositoryImpl.kt:91)
                                                                                                    	at io.homeassistant.companion.android.home.HomePresenterImpl.getAllTileShortcuts(HomePresenterImpl.kt:237)
                                                                                                    	at io.homeassistant.companion.android.home.MainViewModel$loadShortcutTileEntities$1.invokeSuspend(MainViewModel.kt:159)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
                                                                                                    	at kotlinx.coroutines.EventLoop.processUnconfinedEvent(EventLoop.common.kt:68)
                                                                                                    	at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:375)
                                                                                                    	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:30)
                                                                                                    	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:25)
                                                                                                    	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:110)
                                                                                                    	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)
                                                                                                    	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:56)
                                                                                                    	at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
                                                                                                    	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
                                                                                                    	at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
                                                                                                    	at io.homeassistant.companion.android.home.MainViewModel.loadSettings(MainViewModel.kt:136)
                                                                                                    	at io.homeassistant.companion.android.home.MainViewModel.init(MainViewModel.kt:72)
                                                                                                    	at io.homeassistant.companion.android.home.HomeActivity.onCreate(HomeActivity.kt:49)
                                                                                                    	at android.app.Activity.performCreate(Activity.java:8033)
                                                                                                    	at android.app.Activity.performCreate(Activity.java:8013)
                                                                                                    	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309)
                                                                                                    	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3429)
                                                                                                    	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3612)
                                                                                                    	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
                                                                                                    	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2073)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:106)
                                                                                                    	at android.os.Looper.loop(Looper.java:246)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7690)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:593)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:995)
                                                                                                    	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@a540fb0, Dispatchers.Main.immediate]
```

Reply:
Ohh, sorry...

It should be fixed in c298bb01

Reply:
I can confirm the crash is resolved, but if you end up in a state like this (after updating, a second tile is added before the first one is updated) with one tile having ID `null` and another a valid value, the old tile/`null` shortcuts will never be retrieved again.

Not sure how likely it is, but it feels like something that should be handled as normally the app pushes an update to the system (infinite validity of tiles) and the system didn't request it after updating the app.

Reply:
Thanks for testing it, and also for this observation.

FYI I'm currently working on refactoring the logic in `WearPrefsRepositoryImpl`. I came up with an idea on how to make it much more clear and also less error-prone by adhering to the Single Source of Truth principle. It will also address your concern.